File: server/src/HallDates.hs
{-# LANGUAGE OverloadedStrings #-}    
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ViewPatterns #-}

module HallDates (
    module HallDates
)   where


import BasicPrelude

import Data.Time
import Data.Time.Calendar.WeekDate (toWeekDate)
import Data.Functor.Identity

class (Monad m) => MonadTime m where
    getTime :: m UTCTime

instance MonadTime IO where
    getTime = getCurrentTime

instance MonadTime Identity where
    getTime = pure $ UTCTime (ModifiedJulianDay 57564) (secondsToDiffTime 82438)

data Message = NoHallToday
             | BreakFastToday
             | LunchToday
             | BrunchToday
             | DinnerToday
             | MCRDinnerToday
             | NoHallForAges
             | MaybeHall
             deriving (Show, Eq, Ord, Read, Enum, Bounded)

interpret :: Message -> Text
interpret NoHallToday    = "Hall is closed today."
interpret BreakFastToday = "Hall is serving breakfast today."
interpret LunchToday     = "Hall is serving lunch today."
interpret BrunchToday    = "Hall is serving brunch today."
interpret DinnerToday    = "Hall is serving dinner today."
interpret MCRDinnerToday = "There’s an MCR dinner tonight!"
interpret NoHallForAges  = "Hall is closed until August 31st."
interpret MaybeHall      = "Who knows. Better get on your bike."

ishallopen :: MonadTime m => m [Message]
ishallopen = do
        currentTime <- getTime
        let today = utctDay currentTime
        pure $ ishallopentoday today

ishallopentoday :: Day -> [Message]
ishallopentoday (subtract 57564 . toModifiedJulianDay -> day)
    | day >= 95 || day < 0  = [MaybeHall]
    | day < 48 || day >= 66 = 
        case day `mod` 7 of
            0 -> [NoHallToday]
            1 -> [DinnerToday]
            _ -> [BreakFastToday, LunchToday]
    | day >= 48 && day < 67 = [NoHallForAges]
    | otherwise             = [MaybeHall]

File: halldates.js
var then = moment("2016-06-25");
var now = moment();
var today = now.diff(then, 'days');

function getMessage(day) {
    if (day === 17) {
        return 'There’s an MCR dinner tonight!'
    } else if (day >= 96 || day < 0) {
        return 'Who knows. Better get on your bike.'
    } else if (day < 48 || day >= 67) {
        switch (day % 7) {
            case 0:
                return 'Hall is closed today.'
                break;
            case 1:
                return 'Hall is serving dinner today.'
                break;
            default:
                return 'Hall is serving breakfast and lunch today.';
        }
    } else if (day >= 48 && day < 68) {
        return 'Hall is closed until August 31st.'
    } else {
        return 'Who knows. Better get on your bike.'
    }
};

$.ajax({
    url: "https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.14.1/moment.min.js",
    dataType: "script",
    cache: true
}).done($("#message").html(getMessage(today)));
